### PR TITLE
Use recycled byte arrays for peer recovery

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently, every peer recovery file chunk request must allocated the
bytes for the file chunk when the request is received. This PR 
implements to concept of a releasable bytes reference which certain 
stream types can optimize to be shared or pooled. This commit implements
this mechanism for the file chunk requests. The pooled chunks are 
managed at the transport level and will be released when the response
is submitted. Additionally, the releasable references are ref counted
so async implementations can retain them.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Not a proper review, I was just curious about this change, but I left one small comment.

File: server/src/main/java/org/elasticsearch/transport/InboundMessage.java [hunk 1]
```diff
@@ -18,21 +18,34 @@
...
 
 import java.io.Closeable;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 
 public abstract class InboundMessage extends NetworkMessage implements Closeable {
 
     private final StreamInput streamInput;
+    private final List<Releasable> managedResources = new ArrayList<>(4);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundMessage.java]: this is unused

Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundMessage.java]: this is unused

File: server/src/main/java/org/elasticsearch/transport/InboundMessage.java [hunk 2]
```diff
@@ -98,6 +120,7 @@ InboundMessage deserialize(BytesReference reference) throws IOException {
                 return message;
             } finally {
                 if (success == false) {
+                    Releasables.close(managedResources);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/transport/InboundMessage.java]: Shouldn't the list always be empty here

File: server/src/main/java/org/elasticsearch/transport/TcpTransportChannel.java [hunk 3]
```diff
@@ -69,6 +75,7 @@ public void sendResponse(Exception exception) throws IOException {
         try {
             outboundHandler.sendErrorResponse(version, channel, requestId, action, exception);
         } finally {
+            Releasables.close(toRelease);
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/transport/TcpTransportChannel.java]: Also note that the closing logic currently relies on `InboundHandler.messageReceived` not throwing any exception. Given that the initial lifecycle (until `retain()` is called) is ultimately controlled through that method, I wonder if we need extra safekeeping.

**Review by ywelsch — COMMENTED** [review]
I've left some initial comments. I think that the overall approach could work, but also think that getting this right without leaking / incorrectly accessing stuff is going to be tricky. We should add as many safeguards as possible.

**Comment by tbrooks8** [comment]
This work will probably come back at some point. But this PR is pretty out of date. Closing.

**Pull request closed without merging**